tss-esapi/src/interface_types/algorithm.rs

```diff
@@ -0,0 +1,466 @@
+            HashingAlgorithm::Sm3_256 => AlgorithmIdentifier::Sm3_256,
+            HashingAlgorithm::Sha3_256 => AlgorithmIdentifier::Sha3_256,
+            HashingAlgorithm::Sha3_384 => AlgorithmIdentifier::Sha3_384,
+            HashingAlgorithm::Sha3_512 => AlgorithmIdentifier::Sha3_512,
+            HashingAlgorithm::Null => AlgorithmIdentifier::Null,
+        }
+    }
+}
+
+impl TryFrom<AlgorithmIdentifier> for HashingAlgorithm {
+    type Error = Error;
+
+    fn try_from(algorithm_identifier: AlgorithmIdentifier) -> Result<Self> {
+        match algorithm_identifier {
+            AlgorithmIdentifier::Sha1 => Ok(HashingAlgorithm::Sha1),
+            AlgorithmIdentifier::Sha256 => Ok(HashingAlgorithm::Sha256),
+            AlgorithmIdentifier::Sha384 => Ok(HashingAlgorithm::Sha384),
+            AlgorithmIdentifier::Sha512 => Ok(HashingAlgorithm::Sha512),
+            AlgorithmIdentifier::Sm3_256 => Ok(HashingAlgorithm::Sm3_256),
+            AlgorithmIdentifier::Sha3_256 => Ok(HashingAlgorithm::Sha3_256),
+            AlgorithmIdentifier::Sha3_384 => Ok(HashingAlgorithm::Sha3_384),
+            AlgorithmIdentifier::Sha3_512 => Ok(HashingAlgorithm::Sha3_512),
+            AlgorithmIdentifier::Null => Ok(HashingAlgorithm::Null),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+impl From<HashingAlgorithm> for TPMI_ALG_HASH {
+    fn from(hashing_algorithm: HashingAlgorithm) -> Self {
+        AlgorithmIdentifier::from(hashing_algorithm).into()
+    }
+}
+
+impl TryFrom<TPMI_ALG_HASH> for HashingAlgorithm {
+    type Error = Error;
+
+    fn try_from(tpmi_alg_hash: TPMI_ALG_HASH) -> Result<Self> {
+        HashingAlgorithm::try_from(AlgorithmIdentifier::try_from(tpmi_alg_hash)?)
+    }
+}
+
+/// Enum containing the supported keyed hash scheme
+///
+/// # Details
+/// This corresponds to TPMI_ALG_KEYEDHASH_SCHEME interface type.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum KeyedHashSchemeAlgorithm {
+    Hmac,
+    Xor,
+    Null,
+}
+
+impl From<KeyedHashSchemeAlgorithm> for AlgorithmIdentifier {
+    fn from(keyed_hash_scheme: KeyedHashSchemeAlgorithm) -> Self {
+        match keyed_hash_scheme {
+            KeyedHashSchemeAlgorithm::Hmac => AlgorithmIdentifier::Hmac,
+            KeyedHashSchemeAlgorithm::Xor => AlgorithmIdentifier::Xor,
+            KeyedHashSchemeAlgorithm::Null => AlgorithmIdentifier::Null,
+        }
+    }
+}
+
+impl TryFrom<AlgorithmIdentifier> for KeyedHashSchemeAlgorithm {
+    type Error = Error;
+    fn try_from(algorithm_identifier: AlgorithmIdentifier) -> Result<Self> {
+        match algorithm_identifier {
+            AlgorithmIdentifier::Hmac => Ok(KeyedHashSchemeAlgorithm::Hmac),
+            AlgorithmIdentifier::Xor => Ok(KeyedHashSchemeAlgorithm::Xor),
+            AlgorithmIdentifier::Null => Ok(KeyedHashSchemeAlgorithm::Null),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
```

Comment:
I'm wondering if at some point we should make these error codes more expressive. Right now, whether the first conversion from `TPMI_ALG_KEYEDHASH_SCHEME` fails or this one fails, we get the same `InvalidParam`, right? But this isn't really important for this PR.

It's really nice to have the conversion implemented like this, though, so we don't have duplicate stuff 💯

Reply:
Yes, I liked the the double conversions it makes it more clear what goes wrong. The same goes for AlgorithmIdentifier.